crypto/src/mls/credential/trust_anchor.rs

```diff
@@ -0,0 +1,1647 @@
+};
+
+use crate::{
+    mls::{
+        client::Client,
+        conversation::{group_info::MlsGroupInfoBundle, handshake::MlsCommitBundle, ConversationId},
+        MlsCentral,
+    },
+    prelude::MlsConversation,
+    MlsError,
+};
+use crate::{CryptoError, CryptoResult};
+
+/// A wrapper containing the configuration for trust anchors to be added in the group's context
+/// extensions
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PerDomainTrustAnchor {
+    /// Domain name in which the trust anchor belongs to
+    pub domain_name: String,
+    /// PEM encoded certificate chain
+    pub intermediate_certificate_chain: String,
+}
+
+impl TryFrom<&openmls::extensions::PerDomainTrustAnchor> for PerDomainTrustAnchor {
+    type Error = CryptoError;
+
+    fn try_from(value: &openmls::extensions::PerDomainTrustAnchor) -> Result<Self, Self::Error> {
+        let pems: Vec<_> = value
+            .certificate_chain()
+            .iter()
+            .map(|cert| pem::Pem::new("CERTIFICATE", cert.clone()))
+            .collect();
+        Ok(Self {
+            domain_name: String::from_utf8(value.domain_name().to_vec())?,
+            intermediate_certificate_chain: pem::encode_many(&pems),
+        })
+    }
+}
+
+impl PerDomainTrustAnchor {
+    /// Converts to the OpenMls's counterpart of this struct
+    /// It performs validation first
+    pub fn into_mls(
+        self,
+        backend: &MlsCryptoProvider,
+        group_context: Option<&GroupContext>,
+    ) -> CryptoResult<openmls::extensions::PerDomainTrustAnchor> {
+        let certificate_chain = self.validate(backend, group_context)?;
+        Ok(openmls::extensions::PerDomainTrustAnchor::new(
+            self.domain_name.into(),
+            openmls::prelude::CredentialType::X509,
+            certificate_chain,
+        )
+        .map_err(MlsError::from)?)
+    }
+}
+
+impl PerDomainTrustAnchor {
+    /// Validates the trust anchor and return its encoded chain encoded to der.
+    fn validate(
+        &self,
+        backend: &MlsCryptoProvider,
+        group_context: Option<&GroupContext>,
+    ) -> CryptoResult<Vec<Vec<u8>>> {
+        // parse PEM
+        let certificate_chain = pem::parse_many(&self.intermediate_certificate_chain)?
+            .iter()
+            .map(|p| Certificate::from_der(p.contents()))
+            .collect::<Result<PkiPath, x509_cert::der::Error>>()?;
+
+        // verify domain_name is unique
+        if let Some(group_context) = group_context {
+            if group_context
+                .extensions()
+                .per_domain_trust_anchors()
+                .is_some_and(|anchors| {
+                    anchors
+                        .iter()
+                        .any(|anchor| String::from_utf8_lossy(anchor.domain_name()) == self.domain_name)
+                })
+            {
+                return Err(CryptoError::DuplicateDomainName);
+            }
+        }
+
+        // self signed or incomplete chains are not allowed
```

Comment:
This comment is misleading

```suggestion
        // empty chains are not allowed
```